src/Microsoft.AspNetCore.ResponseCaching/Internal/ResponseCacheStream.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.ResponseCaching.Internal
+{
+    internal class ResponseCacheStream : Stream
+    {
+        private readonly Stream _innerStream;
+
+        public ResponseCacheStream(Stream innerStream)
+        {
+            _innerStream = innerStream;
+        }
+
+        public MemoryStream BufferedStream { get; } = new MemoryStream();
+
+        public override bool CanRead => _innerStream.CanRead;
+
+        public override bool CanSeek => _innerStream.CanSeek;
+
+        public override bool CanWrite => _innerStream.CanWrite;
+
+        public override long Length => _innerStream.Length;
+
+        public override long Position
+        {
+            get { return _innerStream.Position; }
+            set { _innerStream.Position = value; }
+        }
+
+        public override void SetLength(long value)
+        {
+            BufferedStream.SetLength(value);
+            _innerStream.SetLength(value);
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            BufferedStream.Seek(offset, origin);
+            return _innerStream.Seek(offset, origin);
+        }
+
+        public override void Flush()
+            => _innerStream.Flush();
+
+        public override async Task FlushAsync(CancellationToken cancellationToken)
+            => await _innerStream.FlushAsync();
+
+        // Underlying stream is write-only, no need to override other read related methods
+        public override int Read(byte[] buffer, int offset, int count)
+            => _innerStream.Read(buffer, offset, count);
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            BufferedStream.Write(buffer, offset, count);
+            _innerStream.Write(buffer, offset, count);
+        }
+
+        public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            await BufferedStream.WriteAsync(buffer, offset, count, cancellationToken);
+            await _innerStream.WriteAsync(buffer, offset, count, cancellationToken);
+        }
+
+        public override void WriteByte(byte value)
+        {
+            BufferedStream.WriteByte(value);
```

Comment:
This is still fully buffering every response. When OnResponseStarting fires and it calculates CacheResponse, if it evaluates to false it should tell ResponseCacheStream to stop buffering and discard any already buffered data.